### ISSUE TITLE
compare and swap refactor

### ISSUE THREAD

**Issue by spacejam** [spacejam/sled#820]
- [ ] rename cas to compare_and_swap
- [ ] keep a cas method that just does a compiler error with a message to use compare_and_swap, which will have a different error signature (if it had the same signature we could do a deprecation, but because it changes, we should give people an error message that includes a short migration guide)
- [ ] add CompareAndSwapError that includes current and proposed-replacement values
- [ ] improve docs, especially around failures

**Comment by sitano** [comment]
Trying to imagine a workaround for not being able to store Key and Values ref types in an Error Result struct due to the lack of generics in this enum and lifetime:
```
pub enum Error {
    ...
    /// A compare and swap failure
    CompareAndSwap {
        key: &[u8], <---------------------- require a lifetime which cant be passed here
        old: Option<&[u8]>,
        new: Option<crate::ivec::IVec>,
    },
}
```

**Comment by spacejam** [comment]
closed by #832
